pkg/reconciler/apibinding/apibinding_controller.go

```diff
@@ -0,0 +1,347 @@
+	apislisters "github.com/kcp-dev/kcp/pkg/client/listers/apis/v1alpha1"
+)
+
+const (
+	controllerName                    = "kcp-apibinding"
+	indexAPIBindingsByWorkspaceExport = "apiBindingsByWorkspaceExport"
+	//indexAPIResourceSchemasByCRD       = "apiResourceSchemasByCRD"
+	indexAPIExportsByAPIResourceSchema = "apiExportsByAPIResourceSchema"
+	shadowWorkspaceName                = "system:bound-crds"
+)
+
+// NewController returns a new controller for APIBindings.
+func NewController(
+	crdClient apiextensionclientset.ClusterInterface,
+	kcpClient kcpclient.ClusterInterface,
+	apiBindingInformer apisinformers.APIBindingInformer,
+	apiExportInformer apisinformers.APIExportInformer,
+	apiResourceSchemaInformer apisinformers.APIResourceSchemaInformer,
+	crdInformer apiextensionsinformers.CustomResourceDefinitionInformer,
+) (*controller, error) {
+	queue := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), controllerName)
+
+	c := &controller{
+		queue:                    queue,
+		crdClusterClient:         crdClient,
+		kcpClusterClient:         kcpClient,
+		apiBindingsLister:        apiBindingInformer.Lister(),
+		apiBindingsIndexer:       apiBindingInformer.Informer().GetIndexer(),
+		apiExportsLister:         apiExportInformer.Lister(),
+		apiExportsIndexer:        apiExportInformer.Informer().GetIndexer(),
+		apiResourceSchemaLister:  apiResourceSchemaInformer.Lister(),
+		apiResourceSchemaIndexer: apiResourceSchemaInformer.Informer().GetIndexer(),
+		crdLister:                crdInformer.Lister(),
+		crdIndexer:               crdInformer.Informer().GetIndexer(),
+		deletedCRDTracker:        newLockedStringSet(),
+	}
+
+	apiBindingInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { c.enqueueAPIBinding(obj) },
+		UpdateFunc: func(_, obj interface{}) { c.enqueueAPIBinding(obj) },
+		DeleteFunc: func(obj interface{}) { c.enqueueAPIBinding(obj) },
+	})
+
+	if err := apiBindingInformer.Informer().AddIndexers(cache.Indexers{
+		indexAPIBindingsByWorkspaceExport: indexAPIBindingByWorkspaceExport,
+	}); err != nil {
+		return nil, err
+	}
+
+	crdInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: func(obj interface{}) bool {
+			crd, ok := obj.(*apiextensionsv1.CustomResourceDefinition)
+			if !ok {
+				return false
+			}
+
+			return crd.ClusterName == shadowWorkspaceName
+		},
+		Handler: cache.ResourceEventHandlerFuncs{
+			AddFunc:    func(obj interface{}) { c.enqueueCRD(obj) },
+			UpdateFunc: func(_, obj interface{}) { c.enqueueCRD(obj) },
+			DeleteFunc: func(obj interface{}) {
+				meta, err := meta.Accessor(obj)
+				if err != nil {
+					runtime.HandleError(err)
+					return
+				}
+
+				// If something deletes one of our bound CRDs, we need to keep track of it so when we're reconciling,
+				// we know we need to recreate it.
+				c.deletedCRDTracker.Add(meta.GetName())
```

Comment:
who will recreate them? We on resync?

Reply:
Yeah, this handled a race I was seeing where I manually deleted the CRD, this sync got triggered, and it was still seeing the CRD in the cache, so it didn't end up recreating the CRD. With this tracker in place, the recreates worked.

Reply:
ack

Reply:
mention that in the comment